ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/math/Greatest.java

```diff
@@ -0,0 +1,113 @@
+import io.confluent.ksql.function.FunctionCategory;
+import io.confluent.ksql.function.udf.Udf;
+import io.confluent.ksql.function.udf.UdfDescription;
+import io.confluent.ksql.function.udf.UdfParameter;
+import io.confluent.ksql.function.udf.UdfSchemaProvider;
+import io.confluent.ksql.schema.ksql.SqlArgument;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.util.DecimalUtil;
+import io.confluent.ksql.util.KsqlConstants;
+import java.math.BigDecimal;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+@UdfDescription(
+    name = "greatest",
+    category = FunctionCategory.MATHEMATICAL,
+    description = "Returns the highest value among a variable number of consistently typed columns",
+    author = KsqlConstants.CONFLUENT_AUTHOR
+)
+public class Greatest {
+
+  @Udf
+  public Double greatest(@UdfParameter final Double val, @UdfParameter final Double... vals) {
+
+    final Stream<Double> toConcat = Stream.of(val);
+
+    return Streams.concat(toConcat, Arrays.stream(vals))
+        .filter(Objects::nonNull)
+        .max(Double::compareTo)
+        .orElse(null);
+  }
+
+  @Udf
+  public Integer greatest(@UdfParameter final Integer val, @UdfParameter final Integer... vals) {
+    final Stream<Integer> toConcat = Stream.of(val);
+    return Stream.concat(toConcat, Arrays.stream(vals))
+        .filter(Objects::nonNull)
+        .max(Integer::compareTo)
+        .orElse(null);
+  }
+
+  @Udf
+  public Long greatest(@UdfParameter final Long val, @UdfParameter final Long... vals) {
+    final Stream<Long> toConcat = Stream.of(val);
+    return Stream.concat(toConcat, Arrays.stream(vals))
+        .filter(Objects::nonNull)
+        .max(Long::compareTo)
+        .orElse(null);
+  }
+
+  @Udf
+  public String greatest(@UdfParameter final String val, @UdfParameter final String... vals) {
+
+    final Stream<String> toConcat = Stream.of(val);
+
+    return Streams.concat(toConcat, Arrays.stream(vals))
+        .filter(Objects::nonNull)
+        .max(String::compareTo)
+        .orElse(null);
+  }
+
+  @Udf(schemaProvider = "greatestDecimalProvider")
+  public BigDecimal greatest(@UdfParameter final BigDecimal val,
+      @UdfParameter final BigDecimal... vals) {
+
+    final Stream<BigDecimal> toConcat = Stream.of(val);
+
+    return Streams.concat(toConcat, Arrays.stream(vals))
+        .filter(Objects::nonNull)
+        .max(Comparator.naturalOrder())
+        .orElse(null);
+  }
+
+  @UdfSchemaProvider
+  public SqlType leastDecimalProvider(final List<SqlArgument> params) {
+
+    SqlType wideDecimal = params.stream()
+        .filter(Objects::nonNull)
+        .findAny()
+        .get()
+        .getSqlTypeOrThrow();
+
+    for (SqlArgument s : params) {
+      if (s.getSqlType().isPresent()) {
+        wideDecimal = DecimalUtil.widen(s.getSqlTypeOrThrow(), wideDecimal);
+      }
+    }
+
+    return wideDecimal;
+  }
+
+}
```

Comment:
-Is there a cleaner way to find the "widest" decimal scale/precision of the numbers while checking for null `SqlArgument`s?

As I'm writing this I realize the stream on line 98 might not be doing what I want it to, I'll double check this tomorrow but hopefully the intent is clear. Regardless I'd look for a way to write it so we can get the result without iterating twice.

Reply:
https://docs.oracle.com/javase/tutorial/collections/streams/reduction.html#reduce